Blog App - Processing data in models

gitbygaurav opened the pull request:
### **Hi CR 👋**

![AidanHelloAidanGallagherHelloGIF](https://user-images.githubusercontent.com/68365169/228588690-5941a9ff-bb22-45d4-a2bb-cb25efffca35.gif)

**In this branch, I have done the following changes :-**

- [x] Created model classes for all entities.
- [x] Set up associations between models.
- [x] Created 2 users
- [x] Created 4 posts written by one of the users.
- [x] Created 6 posts comments for one of the posts created
- [x] In the User model, I created a method that returns the 3 most recent posts for a given user.
- [x] In the Post model, I created a method that updates the posts counter for a user.
- [x] In the Post model, I created a method that returns the 5 most recent comments for a given post.
- [x] In the Comment model, I created a method that updates the comments counter for a post.
- [x] In the Like model, I created a method that updates the likes counter for a post.

Please review my changes
Thanks🍻

gitbygaurav committed:
process data in models done

gitbygaurav committed:
Fix linter errors

CollinsTatang reviewed (commented):
Hi @gauravgangwarweb , 

While you made a great effort in this project. Unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

You have submitted a project, that does not pass the linter test.

- [ ] Please your [Linters](https://github.com/microverseinc/linters-config/tree/master/ror) for this project should be properly setup.
- [ ] Your linter result should be green, this means that the last commit has a green check mark (✅).
- [ ] Please there should be no custom changes to your linter config.


Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

gitbygaurav committed:
Update README.md

gitbygaurav committed:
Fix linters

gitbygaurav committed:
Merge branch 'feature-processing-data' of github.com:gauravgangwarweb/Blog-App into feature-processing-data

gitbygaurav committed:
Fix linter setup

gitbygaurav committed:
Fix github workflow file

gitbygaurav commented:
Hi @CollinsTatang 
I have added the linters correctly. I am not getting that why the linters are not running on GitHub. Can you please give me some hints.

gitbygaurav committed:
Add some changes

CollinsTatang commented:
> Hi @CollinsTatang I have added the linters correctly. I am not getting that why the linters are not running on GitHub. Can you please give me some hints.

Please follow the set up link I added above @gauravgangwarweb

gitbygaurav committed:
Fix github workflows

gitbygaurav commented:
> > Hi @CollinsTatang I have added the linters correctly. I am not getting that why the linters are not running on GitHub. Can you please give me some hints.
> 
> Please follow the set up link I added above @gauravgangwarweb

I have followed every step of that then also it is not working. @CollinsTatang

mukhammadrizooff commented on app/models/user.rb:
```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  has_many :posts
+  has_many :comments
+  has_many :likes
```
- [ ] [REQUIRED] Good job so far! The `associations` between `models` should be set up correctly according to the given [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd.png), kindly add also required `foreign_key`. Please implement my suggestions below 🙏 

```ruby
# app/models/user.rb
has_many :posts, foreign_key: 'author_id'
has_many :comments, foreign_key: 'author_id'
has_many :likes, foreign_key: 'author_id'
```

mukhammadrizooff commented on app/models/post.rb:
```diff
@@ -0,0 +1,13 @@
+class Post < ApplicationRecord
+  belongs_to :user
+  has_many :likes
+  has_many :comments
```
- [ ] [REQUIRED] Good job so far! The `associations` between `models` should be set up correctly according to the given [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd.png). Please implement my suggestions below 🙏 

```ruby
# app/models/post.rb
belongs_to :author, class_name: 'User'
has_many :comments
has_many :likes
```

mukhammadrizooff commented on app/models/comment.rb:
```diff
@@ -0,0 +1,8 @@
+class Comment < ApplicationRecord
+  belongs_to :user
+  belongs_to :post
```
- [ ] [REQUIRED] Good job so far! The `associations` between `models` should be set up correctly according to the given [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd.png). Please implement my suggestions below 🙏 

```ruby
# app/models/comment.rb
belongs_to :author, class_name: 'User'
belongs_to :post
```

mukhammadrizooff commented on app/models/like.rb:
```diff
@@ -0,0 +1,8 @@
+class Like < ApplicationRecord
+  belongs_to :user
+  belongs_to :post
```
- [ ] [REQUIRED] Good job so far! The `associations` between `models` should be set up correctly according to the given [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd.png). Please implement my suggestions below 🙏 

```ruby
# app/models/like.rb
belongs_to :author, class_name: 'User'
belongs_to :post
```

mukhammadrizooff reviewed (changes requested):
Hi @gauravgangwarweb, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

<img src="https://media2.giphy.com/media/Jq824R93JsLwZCaiSL/giphy.gif"/>

## Highlights
- Correct GitFlow ✔️ 
- No linter errors ✔️ 
- Good implementations ✔️ 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

CollinsTatang commented:
> > > Hi @CollinsTatang I have added the linters correctly. I am not getting that why the linters are not running on GitHub. Can you please give me some hints.
> > 
> > 
> > Please follow the set up link I added above @gauravgangwarweb
> 
> I have followed every step of that then also it is not working. @CollinsTatang

Sorry, it was a general issue but it has been solved now I see your linter passing.

gitbygaurav committed:
Add required changes

rotimiazeez commented on app/models/post.rb:
```diff
@@ -0,0 +1,13 @@
+  belongs_to :author, class_name: 'User'
+  has_many :comments
+  has_many :likes
+
```
-[ OPTIONAL ] It will be nice if you can use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the #update_post_counter method every time a new post has been saved to database, then make the update_post_counter method private as seen below:
```
  # app/models/post.rb
  after_save :update_post_counter

  private

  def update_post_counter
    author.increment!(:posts_counter)
  end
```

rotimiazeez commented on app/models/like.rb:
```diff
@@ -0,0 +1,8 @@
+class Like < ApplicationRecord
+  belongs_to :author, class_name: 'User'
+  belongs_to :post
+
```
-[ OPTIONAL ] It will be nice if you can use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the #update_like_counter method every time a new like has been saved to database, then make the update_like_counter method private as seen below:
```
  # app/models/like.rb
  after_save :update_like_counter

  private

  def update_like_counter
    author.increment!(:likes_counter)
  end
```

rotimiazeez commented on app/models/comment.rb:
```diff
@@ -0,0 +1,8 @@
+class Comment < ApplicationRecord
+  belongs_to :author, class_name: 'User'
+  belongs_to :post
+
```
-[ OPTIONAL ] It will be nice if you can use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the #update_comment_counter method every time a new comment has been saved to database, then make the update_comment_counter method private as seen below:
```
  # app/models/comment.rb
  after_save :update_comment_counter

  private

  def update_comment_counter
    author.increment!(:comments_counter)
  end
```

rotimiazeez reviewed (approved):
Hi @gauravgangwarweb 👏🏽, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlight
- Good job working on the Suggested changes 👌🏽
- No linter error ✔️

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code. Some of them were simply missed by the previous reviewer, and addressing them will improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me(@rotimiazeez) in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561), you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

gitbygaurav committed:
Add required optional chnages

gitbygaurav committed:
Fix linter errors

gitbygaurav commented:
Hii @rotimiazeez 
Thanks for approving and giving me some optional changes I have implemented those changes.
